<a href="https://colab.research.google.com/github/AIPracticeUser/Mathematical-Foundations/blob/main/Maths_2_(Scalars%2C_Vectors%2C_Matrices).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Scalars (Rank 0 Tensors)

In [1]:
x = 25
x

25

In [2]:
type(x) #If want to use specific like int16, need Numpy 

int

In [3]:
y = 3

In [4]:
py_sum = x + y

In [5]:
type(py_sum)

int

In [6]:
x_float = 25.0
float_sum = x_float + y
float_sum

28.0

In [7]:
type(float_sum)

float

## Scalars in PyTorch

- PyTorch and TensorFlow are the two most popular automatic differentiation librarie in Python, itself the most popular programming language in ML
- PyTorch tensors are designed to be pythonic, i.e., to feel and behave like NumPy arrays
- The advantage of PyTorch tensors relative to NumPy arrays is that they easily be used for operations on GPU 


In [8]:
import torch

In [9]:
x_pt = torch.tensor(25, dtype=torch.float16) #type specification optional, e.g.: dtype=torch.float16
x_pt

tensor(25., dtype=torch.float16)

In [10]:
x_pt.shape #Scalars has no dimensions

torch.Size([])

# Scalars in TensorFlow 
Tensors created with a wrapper
Documentation : https://www.tensorflow.org/guide/tensor

- tf.Variable
- tf.constant
- tf.placeholder
- tf.SparseTensor
Most widely-used is tf.Variable, which we'll use here.

As with TF tensors, in PyTorch we can similarly perform operations, and we can easily convert to and from NumPy arrays

Full list of tensor data types: https://www.tensorflow.org/api_docs/python/tf/dtypes/DType

In [11]:
import tensorflow as tf

In [12]:
x_tf = tf.Variable(25, dtype=tf.int16) #dtype is optional
x_tf

<tf.Variable 'Variable:0' shape=() dtype=int16, numpy=25>

In [13]:
x_tf.shape #Also shows 0 dimensions

TensorShape([])

In [14]:
y_tf = tf.Variable(3, dtype=tf.int16)

In [15]:
x_tf + y_tf

<tf.Tensor: shape=(), dtype=int16, numpy=28>

In [16]:
tf_sum = tf.add(x_tf, y_tf) #Same operation as above
tf_sum

<tf.Tensor: shape=(), dtype=int16, numpy=28>

In [17]:
tf_sum.numpy() # note that Numpy operations automatically convert tensors to Numpy arrays and vice versas

28

In [18]:
type(tf_sum.numpy())

numpy.int16

In [19]:
tf_float = tf.Variable(25, dtype=tf.float16)
tf_float

<tf.Variable 'Variable:0' shape=() dtype=float16, numpy=25.0>

## Vectors (Rank 1 Tensors) in NumPy

In [20]:
import numpy as np

In [21]:
x = np.array([25, 2, 5]) #type argument is optional, eg: dtype=np.float16
x

array([25,  2,  5])

In [22]:
type(x)

numpy.ndarray

In [23]:
type(x[0]) #all individual numbers in the array is int64

numpy.int64

In [24]:
x = np.array([25, 2, 5], dtype=np.float16)
x

array([25.,  2.,  5.], dtype=float16)

In [25]:
type(x[0]) #change to float16

numpy.float16

In [26]:
len(x)

3

In [27]:
x.shape

(3,)

## Vector Transposition

In [28]:
# Transposing a regular 1-D array has no effect
x_t = x.T
x_t

array([25.,  2.,  5.], dtype=float16)

In [29]:
x_t.shape

(3,)

In [30]:
# however, if we use nested matrix-style brackets, it chnages
y = np.array([[25,2,5]])
y

array([[25,  2,  5]])

In [31]:
y.shape # it becomes a 2-dimensional

(1, 3)

In [32]:
# can transpose a matrix with a dimension of length 1, which is mathemtically eqivalent:
y_t = y.T
y_t

array([[25],
       [ 2],
       [ 5]])

In [33]:
y_t.shape #this is now a column vector as it has 3 rows and 1 column

(3, 1)

In [34]:
# Reverse back to its original form
y_t.T

array([[25,  2,  5]])

In [35]:
y_t.T.shape

(1, 3)

## Zero Vectors

Have no effect if added to another vector

In [36]:
z = np.zeros(3)
z

array([0., 0., 0.])

## Vectors in PyTorch and TensorFlow

In [37]:
x_pt = torch.tensor([25,2,5]) #PyTorch
x_pt

tensor([25,  2,  5])

In [38]:
x_tf = tf.Variable([25,2,5]) #TensorFlow
x_tf

<tf.Variable 'Variable:0' shape=(3,) dtype=int32, numpy=array([25,  2,  5], dtype=int32)>

##$L^2$ Norm

In [39]:
x = np.array([25, 2, 5]) #type argument is optional, eg: dtype=np.float16
x

array([25,  2,  5])

In [40]:
#Applying Forumula
#1. Square all elements in the vectors
#2. Sum them up
#3. Square of the sum (represent by **(1/2))
(25**2 + 2** 2 + 5**2) ** (1/2)

25.573423705088842

This gives us a Euclidean distance of 25.57 from origin [0, 0, 0] to [25,2,5] coordinates 

In [41]:
#Instead of writing the equation, can use numpy built in library to calculate
np.linalg.norm(x)

25.573423705088842

So, it units in this 3-dimensional vector space are meters, then the vector $x$ has a length of 25.6m

$L^1$ Norm
- Used whenever difference between zero and non-zero is key

In [42]:
x

array([25,  2,  5])

In [43]:
np.abs(25) + np.abs(2) + np.abs(5) #Difference result from L1 norm against L2 norm

32

Squared $L^2$ Norm

In [44]:
x

array([25,  2,  5])

In [45]:
#Same as L2 norm forumla but without the square root
(25**2 + 2** 2 + 5**2)

654

Tensor multiplication $(x^tx)$is equal to L2 norm

In [46]:
np.dot(x, x)

654

Max Norm

In [47]:
x

array([25,  2,  5])

In [48]:
#Apply the forum of Max Norm
#1. Take the absolute value of each element in the array
#2. Take the Max value found, 25 in this case
np.max([np.abs(25), np.abs(2), np.abs(5)])

25

# Basis Vectors
- Can be scaled to represent any vector in a given vector space
- Typically use unit vectors along axes of vector space

#Orthogonal Vectors
- x and y are orthogonal vectors if $x^Ty=0$
- 90 degrees angle to each other (assuming non-zero norms)
- n-dimensional space has max n mutally orthogonal vectors

In [49]:
i = np.array([1,0])
i

array([1, 0])

In [50]:
j = np.array([0,1])
j

array([0, 1])

In [51]:
np.dot(i,j) # Dot Operation of Numpy

0

# Matrices (Rank 2 Tensors) in Numpy
- Two-dimensional array of numbers
- if X has 3 rows and 2 columns, shape is (3,2)

In [52]:
# Use array() with nested brackets:
X = np.array([[25,2], [5,26], [3,7]])
X

array([[25,  2],
       [ 5, 26],
       [ 3,  7]])

In [53]:
X.shape #3 rows, 2 columns

(3, 2)

In [54]:
X.size #no of element in the array

6

In [55]:
# Select left colum of matrix X (zero-indexed)
X[:,0]

array([25,  5,  3])

In [56]:
# Select middle row of matrix X
X[1,:]

array([ 5, 26])

In [57]:
#Another slicing-by-index example:
X[0:2 , 0:2]

array([[25,  2],
       [ 5, 26]])

In [58]:
X[0:2]

array([[25,  2],
       [ 5, 26]])

Matrices in PyTorch

In [59]:
X_pt = torch.tensor([[25,2], [5,26], [3,7]])
X_pt

tensor([[25,  2],
        [ 5, 26],
        [ 3,  7]])

In [60]:
X_pt.shape

torch.Size([3, 2])

In [61]:
X_pt[1,:]

tensor([ 5, 26])

Matrices in TensorFlow

In [62]:
X_tf = tf.Variable([[25,2],[5,26],[3,7]])
X_tf

<tf.Variable 'Variable:0' shape=(3, 2) dtype=int32, numpy=
array([[25,  2],
       [ 5, 26],
       [ 3,  7]], dtype=int32)>

In [63]:
tf.rank(X_tf) #Numpy = 2 means Rank 2

<tf.Tensor: shape=(), dtype=int32, numpy=2>

In [64]:
tf.shape(X_tf)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([3, 2], dtype=int32)>

In [65]:
X_tf[1,:]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 5, 26], dtype=int32)>

## Higher-Rank Tensors
- Rank 4 tensors are common for images, where each dimension corresponds to:
1. Number of images in training batch, e.g. ,32
2. Image height in pixels, e.g., 28 for MNIST digits
3. Image width in pixel, e.g.,28
4. Number of color channehls, e.g, 3 for full-color images (RGB)

In [67]:
#Pytorch
images_pt = torch.zeros([32,28,28,3])
images_pt

tensor([[[[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         ...,

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]]],


        [[[0., 0.

In [69]:
images_tf = tf.zeros([32,28,28,3])
images_tf

<tf.Tensor: shape=(32, 28, 28, 3), dtype=float32, numpy=
array([[[[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        ...,

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]],


   